### PR TITLE
Enhance/5450 twg tag infra

### PR DESCRIPTION
**Pull request by kuasha420**
## Summary

Addresses issue:

- #5450 

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: tests/phpunit/integration/Modules/Thank_With_GoogleTest.php [hunk 1]
```diff
@@ -39,32 +46,123 @@ public function set_up() {
...
+		// Snippet should not be present on non singular pages.
+		$this->assertEquals( $output, $mock_content );
+
+		$post_ID = $this->factory()->post->create();
+		$this->go_to( get_permalink( $post_ID ) );
+
+		$output = apply_filters( 'the_content', '<p>Hello World</p>' );
+
+		$this->assertStringContainsString( 'Thank with Google snippet added by Site Kit', $output );
+		$this->assertStringContainsString( '<button twg-button', $output );
```
Comment (eugene-manuilov) [inline tests/phpunit/integration/Modules/Thank_With_GoogleTest.php:141]: This should be moved into a new test file for the Web_Tag class and more tests should be added to cover all three use cases.

File: includes/Modules/Thank_With_Google/Tag_Interface.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
+<?php
```
[the remark this answers is not in the document]
Comment (kuasha420) [reply includes/Modules/Thank_With_Google/Tag_Interface.php:1]: I don't actually know! I saw it used in the `Analytics` module and instinctively followed it. It doesn't appear to add any value so will get rid of it.

File: includes/Modules/Thank_With_Google/Web_Tag.php [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	/**
+	 * Updates the content of the post.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $content Content of the post.
+	 * @return string Content of the post.
+	 */
+	protected function update_the_content( $content ) {
+		if ( ! $this->is_singular_button_post_type_entity() || substr( $this->button_placement, 0, 7 ) !== 'static_' ) { // strlen( 'static_' ) is 7.
```
[the remark this answers is not in the document]
Comment (kuasha420) [reply includes/Modules/Thank_With_Google/Web_Tag.php:172]: Can you elaborate? We are only checking if `substr( $this->button_placement, 0, 7 ) !== 'static_'` for the dynamic ones and only once here. I don't think there's value in extracting it.

File: includes/Modules/Thank_With_Google/Web_Tag.php [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+			return $content;
+		}
+
+		$button_placeholder = "<div counter-button style=\"height: 34px; visibility: hidden; box-sizing: content-box; padding: 12px 0; display: inline-block; overflow: hidden;\"></div>\n<button twg-button style=\"height: 42px; visibility: hidden; margin: 12px 0;\"></button>";
```
[the remark this answers is not in the document]
Comment (kuasha420) [reply includes/Modules/Thank_With_Google/Web_Tag.php:176]: Makes sense, however, we can't use mewline `\n` inside single quotes. I have removed it. Should not be an issue, right?

File: includes/Modules/Thank_With_Google/Web_Tag.php [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	/**
+	 * Gets snippet comment end.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Snippet comment end.
+	 */
+	private function snippet_comment_end() {
+		return sprintf( "\n<!-- %s -->\n", esc_html__( 'End Thank with Google snippet added by Site Kit', 'google-site-kit' ) );
+	}
```
[the remark this answers is not in the document]
Comment (kuasha420) [reply includes/Modules/Thank_With_Google/Web_Tag.php:221]: Nice! I like yours much better and will use it.

File: tests/phpunit/integration/Modules/Thank_With_GoogleTest.php (hunk 1 above)
Comment (kuasha420) [reply tests/phpunit/integration/Modules/Thank_With_GoogleTest.php]: I have moved it to separate file and added more tests. However, the tests are failing on PHP 5.6 and/or WP 4.7. Can you help me with it?

File: includes/Modules/Thank_With_Google/Web_Tag.php [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+		return is_singular( $this->button_post_types );
+	}
+
+	/**
+	 * Add snippet comments around the tag.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $code The tag code.
+	 *
```
Comment (eugene-manuilov) [inline includes/Modules/Thank_With_Google/Web_Tag.php:207]:
```suggestion
```

File: includes/Modules/Thank_With_Google/Web_Tag.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply includes/Modules/Thank_With_Google/Web_Tag.php]: Yes, should be fine. Thanks.

File: tests/phpunit/integration/Modules/Thank_With_Google/Web_TagTest.php [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+		);
+		$this->go_to( get_permalink( $post_ID ) );
+	}
+
+	private function capture_output( $callback ) {
+		ob_start();
+		$callback();
+		$output = ob_get_clean();
+		return $output;
+	}
```
Comment (eugene-manuilov) [inline tests/phpunit/integration/Modules/Thank_With_Google/Web_TagTest.php:148]: This method is not needed. There is the `get_echo` function that does the same: https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/includes/utils.php#L351-L355

File: tests/phpunit/integration/Modules/Thank_With_GoogleTest.php (hunk 1 above)
Comment (eugene-manuilov) [reply tests/phpunit/integration/Modules/Thank_With_GoogleTest.php]: @kuasha420, not sure why it doesn't work, maybe try to add `'post_status'  => 'publish',` when you generate a new post?

File: includes/Modules/Thank_With_Google/Web_Tag.php [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+	/**
+	 * Sets the current button placement.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $button_placement Button placement.
+	 */
+	public function set_button_placement( $button_placement ) {
+		$this->button_placement = $button_placement;
+
```
Comment (eugene-manuilov) [inline includes/Modules/Thank_With_Google/Web_Tag.php:72]:
```suggestion
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @kuasha420. One more suggestion and this is good to go. I'll apply it by myself.

File: includes/Modules/Thank_With_Google/Web_Tag.php [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+	/**
+	 * Sets the current publication ID.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $publication_id Publication ID.
+	 */
+	public function set_publication_id( $publication_id ) {
+		$this->publication_id = $publication_id;
+	}
```
Comment (felixarntz) [inline includes/Modules/Thank_With_Google/Web_Tag.php:61]: This method is unnecessary. The publication ID for this tag is the "main" ID to pass into the constructor, so it should rather be looked up from the property in the parent class. It's a bit redundant to set it here too.

Comment (eugene-manuilov) [reply includes/Modules/Thank_With_Google/Web_Tag.php:61]: Good catch. I have created a follow up PR to address this: #5568 5616

**Pull request merged**